riscv/triggers.cc

```diff
@@ -55,11 +55,18 @@ void trigger_t::tdata3_write(processor_t * const proc, const reg_t val) noexcept
   sselect = (sselect_t)((proc->extension_enabled_const('S') && get_field(val, CSR_TEXTRA_SSELECT(xlen)) <= SSELECT_MAXVAL) ? get_field(val, CSR_TEXTRA_SSELECT(xlen)) : SSELECT_IGNORE);
 }
 
+static reg_t tcontrol_value(const state_t * state) {
+  if (state->tcontrol)
+    return state->tcontrol->read();
+  else
+    return 0;
+}
+
 bool trigger_t::common_match(processor_t * const proc, bool use_prev_prv) const noexcept {
   auto state = proc->get_state();
   auto prv = use_prev_prv ? state->prev_prv : state->prv;
   auto v = use_prev_prv ? state->prev_v : state->v;
-  auto m_enabled = get_action() != 0 || (state->tcontrol->read() & CSR_TCONTROL_MTE);
+  auto m_enabled = get_action() != 0 || (tcontrol_value(state) & CSR_TCONTROL_MTE);
```

Comment:
Hmm... I'd say it looks like a bug even to have it. We don't want to inhibit a trigger match because tcontrol isn't implemented at all.

Reply:
Yes, we should test whether ```tcontrol``` is implemented before checking the ```tcontrol.mte```.

From the coding tracing, trigger detection functions always call the ```common_match()``` and ```allow_action()``` together. Since this PR checks the ```tcontrol.mte``` in ```allow_actions()```, it does not need to check the ```tcontrol.mte``` again in ```common_match()```. In other words, removing the checking on ```tcontrol.mte``` from ```common_match()``` will not change any behavior. 

That is:
```
bool trigger_t::common_match(processor_t * const proc, bool use_prev_prv) const noexcept {
  auto state = proc->get_state();
  auto prv = use_prev_prv ? state->prev_prv : state->prv;
  auto v = use_prev_prv ? state->prev_v : state->v;
  return mode_match(prv, v) && textra_match(proc);
}
```
Could you please help verify this?

Reply:
Removing this line caused some debug trigger tests to fail. I don't entirely understand why, which is probably because it's not clear to me exactly what common_match() takes care of.

Reply:
After another code tracing, although ```trap_common_t::detect_trap_match()``` always calls the ```common_match()``` and ```allow_action()```, it turns out they check the tcontrol.mte depending on different privilege modes. Precisely, the ```common_match()``` matches if the _previous_ privilege mode is not M-mode, and the ```allow_action()``` allows firing if the current privilege mode is not M-mode. This might be the root cause of the reported issue in https://github.com/riscv-software-src/riscv-isa-sim/pull/1777.

That is another clarification that needs to be made. The itrigger/etrigger detects the _previous_ privilege mode from which the trap is taken. In such a case, which privilege mode does tcontrol.mte (Solution 2) consider? I have an intuition that Solution 1 and Solution 2 consider the same privilege mode since they both avoid the reentancy problem to the current privilege mode. (The current Spike lets Solution 1 consider the current privilege mode but Solution 2 the _previous_ one. This PR makes Solution 2 consider both _previous_ and current privilege modes.)

We should figure out the root cause of failures before merging the PR. Would you please provide more information about the debug trigger tests?

Reply:
The tests I ran were the OpenOCD ones at https://github.com/riscv-software-src/riscv-tests/tree/master/debug.

The debug spec rules are about the mode an exception is taken *from*. For etrigger that means spike needs to look at the previous mode because the trigger code isn't called until after the privilege has been changed.

I still think allow_action() should just be part of common_match(). It doesn't look like one is ever called without the other. I'll try that and see if it simplifies things.